### ISSUE TITLE
Excited to see FluidSynth wrap project created

### ISSUE THREAD

**Issue by kcgen** [mesonbuild/fluidsynth#1]
@jpakkane ,

Just curious how this is going, and if any help testing or other feedback is needed.

We will be ready to put this to use straight-away in DOSBox Staging.

In terms of a minimal viable wrap, we are only interested in the library, which I believe only has Glib and threading as dependencies.  (The FluidSynth application has many more optional dependencies, given it interfaces with almost all popular audio APIs across each platform).

**Comment by eli-schwartz** [comment]
We will test and merge whatever wrap you provide, it's up to you to implement however much of the fluidsynth project you think is relevant to being built as a subproject wrap.

(I agree sticking to the library seems like the obvious way to go.)

This isn't "going" anywhere on the mesonbuild team's side of things, though, since wraps are contributed by the people who want them -- we are merely curating them. :D 

tl;dr

@dreamer How's it going? Hoping to see good things soon.

**Comment by kcgen** [comment]
@eli-schwartz , thanks for clarifying whose hands this is in and allowing us to limit the scope to the library-only, as requiring the wrap provide all possible build permutations would be a tall order indeed (and likely unnecessary for those using FluidSynth as a raw PCM generator, such as is our use-case).

**Comment by dreamer** [comment]
Yes, everything's working as intended :) I requested this repo a few days ago and am working on the wrap. I just pushed a WIP version to my fork: https://github.com/mesonbuild/fluidsynth/compare/2.1.7...dreamer:2.1.7 (not ready for the review yet).

Meson having the ability to configure files using CMake template format is *very* helpful - without this feature creating a wrapper would be much more difficult.

The only head-scratcher part was upstream using a separate C program to generate large C files (you can look up the generated files in my wrap repo) - I decided to bundle the end result instead of building the program (it's easier, these files are not *that* big, they are text files and change rarely, and upstream had a pretty serious problem with generating these files during cross-compilation - I don't want to deal with such problem).

I'm not sure why upstream generates these two files every time instead of just committing the files in their repo… Unfortunately, tables in these files are filled with results of floating-point calculations, so they can't be e.g. generated via C++ `constexpr` function… Overall it's an interesting problem to be discussed with upstream maintainers :)

Aside of that I get a result conflicting with upstream documentation (build succeeds without gthread dependency, which is unexpected), and need to figure out if we really don't need any of the numerous optional features of FluidSynth.

**Comment by eli-schwartz** [comment]
Cross compilation should not be a problem, the c program generating the static files should simply be marked with the `native` kwarg to ensure it's built to be run on the build machine instead of the cross-compile target. Unless I'm misunderstanding the problem?

**Comment by kcgen** [comment]
Great progress @dreamer; thanks for linking to your working branch.

One addition is:

``` c++
#define WITH_FLOAT
```
Without it, our call to `fluid_synth_nwrite_float(..)` runs the synthesis using the default integer math and then casts them up to `float`s, instead of running the synthesis itself in floating point. 

For optimized or release builds (of FluidSynth), it would be _really nice_ to tack-on `-ffast-math` to its CXX/CFLAGS (atleast for gcc, clang, and icc; not sure what the msvc equivalent is), because we can certainly afford to fall short of IEEE precision requirements at the 32-bit level, given ultimately our 32-bit floats are cast down to 16-bit (which loses orders-of-magnitude more precision than the impact of fast-math).

For my local build that I test DOSBox Staging with, I also disable everything:

``` c++
	-DOpenMP_gomp_LIBRARY="" \
	-DLIB_SUFFIX="" \
	-Denable-alsa=0 \
	-Denable-aufile=0 \
	-Denable-coreaudio=0 \
	-Denable-coremidi=0 \
	-Denable-dbus=0 \
	-Denable-dsound=0 \
	-Denable-ipv6=0 \
	-Denable-jack=0 \
	-Denable-ladspa=0 \
	-Denable-lash=0 \
	-Denable-libinstpatch=0 \
	-Denable-libsndfile=0 \
	-Denable-midishare=0 \
	-Denable-network=0 \
	-Denable-oss=0 \
	-Denable-portaudio=0 \
	-Denable-pulseaudio=0 \
	-Denable-readline=0 \
	-Denable-sdl2=0 \
	-Denable-systemd=0 \
	-Denable-wasapi=0 \
	-Denable-waveout=0 \
	-Denable-winmidi=0 \
```
Actually, I suppose this is how our Linux and macOS release binaries were packaged too, given that set comes from `contrib/static-fluidsynth`.

I'm not sure if it's possible to chop out even more of its bindings and drivers?
```
  'bindings/fluid_cmd.c',
  'bindings/fluid_filerenderer.c',
  'bindings/fluid_ladspa.c',
  'drivers/fluid_adriver.c',
  'drivers/fluid_mdriver.c',
  'midi/fluid_midi.c',
  'midi/fluid_midi_router.c',
```

**Comment by kcgen** [comment]
`ffast-math` also passes all unit-tests:

``` text
      Start  1: test_sample_cache
 1/21 Test  #1: test_sample_cache ...................   Passed    0.01 sec
      Start  2: test_sfont_loading
 2/21 Test  #2: test_sfont_loading ..................   Passed    0.01 sec
      Start  3: test_sample_rate_change
 3/21 Test  #3: test_sample_rate_change .............   Passed    0.00 sec
      Start  4: test_preset_sample_loading
 4/21 Test  #4: test_preset_sample_loading ..........   Passed    0.01 sec
      Start  5: test_preset_pinning
 5/21 Test  #5: test_preset_pinning .................   Passed    0.01 sec
      Start  6: test_bug_635
 6/21 Test  #6: test_bug_635 ........................   Passed    0.02 sec
      Start  7: test_settings_unregister_callback
 7/21 Test  #7: test_settings_unregister_callback ...   Passed    0.00 sec
      Start  8: test_pointer_alignment
 8/21 Test  #8: test_pointer_alignment ..............   Passed    0.05 sec
      Start  9: test_seqbind_unregister
 9/21 Test  #9: test_seqbind_unregister .............   Passed    0.00 sec
      Start 10: test_synth_chorus_reverb
10/21 Test #10: test_synth_chorus_reverb ............   Passed    0.00 sec
      Start 11: test_snprintf
11/21 Test #11: test_snprintf .......................   Passed    0.00 sec
      Start 12: test_synth_process
12/21 Test #12: test_synth_process ..................   Passed    0.00 sec
      Start 13: test_ct2hz
13/21 Test #13: test_ct2hz ..........................   Passed    0.00 sec
      Start 14: test_sample_validate
14/21 Test #14: test_sample_validate ................   Passed    0.00 sec
      Start 15: test_sfont_unloading
15/21 Test #15: test_sfont_unloading ................   Passed    1.68 sec
      Start 16: test_seq_event_queue_sort
16/21 Test #16: test_seq_event_queue_sort ...........   Passed    0.04 sec
      Start 17: test_seq_scale
17/21 Test #17: test_seq_scale ......................   Passed    0.00 sec
      Start 18: test_seq_evt_order
18/21 Test #18: test_seq_evt_order ..................   Passed    0.00 sec
      Start 19: test_seq_event_queue_remove
19/21 Test #19: test_seq_event_queue_remove .........   Passed    0.08 sec
      Start 20: test_jack_obtaining_synth
20/21 Test #20: test_jack_obtaining_synth ...........   Passed    0.00 sec
      Start 21: test_utf8_open
21/21 Test #21: test_utf8_open ......................   Passed    0.00 sec

100% tests passed, 0 tests failed out of 21
```

``` shell
{C|CXX}FLAGS="-Ofast -ffast-math -ftree-slp-vectorize -ftree-vectorize -msse2 \
              -march=native -frename-registers -flto=2 -DNDEBUG -pipe"
```

**Comment by dreamer** [comment]
OK, back to working on this.

@eli-schwartz 
> Cross compilation should not be a problem, the c program generating the static files should simply be marked with the `native` kwarg to ensure (…)

That was it, thank you! https://github.com/dreamer/fluidsynth/commit/eab9ca181292f4c0eae641aa8106a02a8fd3bfb4

There's one more curveball though. The program generates two files with .c extension, but these files are not compilable - they are being `#include`d in the source code instead… This confuses Meson a bit, as when I add my `custom_target` object to the library using these included sources, Meson attempts to compile them… I tried to work around this using `declare_dependency`, but it seems impossible. I ended up using kwarg `build_by_default : true` in my `custom_target` generating the files - **and it works**.

My question is: is this correct? Will this target be run before generated .c files are included? On the first sight there's no explicit dependency between custom target and library being built…

**Comment by eli-schwartz** [comment]
As far as I know this is currently impossible, and there's no mechanism to add `depend_files:` or `depends:` to an `executable()`. It wasn't really considered useful.

The generated files should really be named *.inc or *.h if they're going to be `#include`d like that.

I wouldn't intuitively expect build_by_default to add a dependency rule. Can you try configuring the project from scratch and then telling ninja to build that one executable rather than the default/all target? Because I'm guessing that given sufficient parallelism or randomness in the order of build edges, you will start hitting problems.

**Comment by dreamer** [comment]
I agree that those files should be named differently (`.inc` or maybe `.h`), but I would need to patch the sources (both generator and sources with `#include` lines), so that's not an option :( Will talk to upstream about it.

> Can you try configuring the project from scratch and then telling ninja to build that one executable rather than the default/all target?

Not sure why building executable would cause the problem to manifest (exec generates the .c files, that's all), but I tried anyway.

Generated `build.ninja` has following relevant targets: `src/make_tables` (executable to generate .c files), `src/fluid_conv_tables.c`, `src/fluid_rvoice_dsp_tables.c` (generated files), (and all the rest). I tried `meson --wipe` and building different targets, several times - every time it worked correctly.

Maybe @jpakkane can shed some light here? Is `build_by_default : true` safe in this case?

**Comment by eli-schwartz** [comment]
I mean the executable that contains source files which #include the generated sources.

Edit: apparently what I really mean is "tell ninja to just build the library". Sorry for that.

**Comment by dreamer** [comment]
OK, I successfully managed to make it fail - it worked earlier because I used default target and not .so file directly:
```
$ cd build
$ meson --wipe
$ ninja libfluidsynth.so
[35/37] Compiling C object libfluidsynth.so.p/src_rvoice_fluid_rvoice_dsp.c.o
FAILED: libfluidsynth.so.p/src_rvoice_fluid_rvoice_dsp.c.o 
ccache cc -Ilibfluidsynth.so.p -I. -I.. -Iinclude -I../include -Isrc -I../src -I../src/bindings -I../src/drivers -I../src/midi -I../src/rvoice -I../src/sfloader -I../src/synth -I../src/utils -I/usr/include/glib-2.0 -I/usr/lib64/glib-2.0/include -fdiagnostics-color=always -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -g -fPIC -pthread -MD -MQ libfluidsynth.so.p/src_rvoice_fluid_rvoice_dsp.c.o -MF libfluidsynth.so.p/src_rvoice_fluid_rvoice_dsp.c.o.d -o libfluidsynth.so.p/src_rvoice_fluid_rvoice_dsp.c.o -c ../src/rvoice/fluid_rvoice_dsp.c
../src/rvoice/fluid_rvoice_dsp.c:24:10: fatal error: fluid_rvoice_dsp_tables.c: No such file or directory
   24 | #include "fluid_rvoice_dsp_tables.c"
```
So, the problem is: when generated files have extension `.c`, I have no way to communicate that to `library` target, because Meson attempts to compile them (I have no way to inform Meson: "whoa, wait a moment, I know these are .c files, but they are not to be compiled - they are #included instead").

**Comment by jpakkane** [comment]
> So, the problem is: when generated files have extension .c, I have no way to communicate that to library target, because Meson attempts to compile them

This is the expected behaviour and works like this by design. Files with "source" extension are compiled. Others are not. This simplifies both implementation and usage but unfortunately causes some friction in these cases (which are quite rare). The "correct" solution is to change the generated file extension to something like `.c.inc`. If you absolutely, positively can not change the source code then you can still generate the files with those new extensions but in addition create (with `configure_file`) a `outputname.c` file whose only contents is this:

    #include<outputname.c.inc>

Yes, it's a bit terrible, but it should work.

**Comment by dreamer** [comment]
@jpakkane 
> This is the expected behaviour and works like this by design. Files with "source" extension are compiled. (…)

Oh, I think we all agree about that :) In this case, the generated files are in wrap, so I can't easily change the `#include` line. It seems like I have only two alternatives right now:
1. Bundle generated files in the wrap repo, or
2. Keep `build_by_default : true` and hope the builds won't fail randomly (so far the failure was happening only when I specifically rebuilt the .so file)

Both solutions are… not great.

*edit* ah - I can't configure_file either because the names of generated files are hardcoded :( but maybe I can generate them to a separate dir and do configure_file based on that… I'll try.

**Comment by eli-schwartz** [comment]
Maybe as a stopgap solution... Create a python script that runs the program for generating the .c file, but the python script declares *two* output files. One of those outputs is an empty file named `do_nothing.h` and only it is added as a source file for the library.

So, meson thinks it needs that header, and it forces the custom_target to run in order to create the `do_nothing.h` and incidentally creates the .c file too (which is invisibly used).

You should then be able to add an index of the custom_target output to only add the first .h output file as an input for the library. Then the depfile ensures future updates to the .c file trigger a recompile of the library.

**Comment by dreamer** [comment]
@eli-schwartz yup, this approach works :) I just avoided listing generated .c files altogether and rely only on the generated fake header to propagate the dependency. https://github.com/dreamer/fluidsynth/commit/62d7c9d8c13d952d50a30a89b52af3353d628389

I need to clean the wrap file a bit and soon it should be ready for review.

**Comment by eli-schwartz** [comment]
That will probably prevent ninja clean from cleaning up the unstated files. Would otherwise work...

**Comment by dreamer** [comment]
I just created a PR with initial version of the wrap. I think this issue should be closed and discussion should continue in #2.
